pydantic/_internal/_generics.py

```diff
@@ -0,0 +1,327 @@
+    else:
+        visit_counts_by_ref[type_ref] += 1
+        yield None
+    if token:
+        _visit_counts_context.reset(token)
+
+
+def recursively_defined_type_refs() -> set[str]:
+    return set((_visit_counts_context.get() or {}).keys())
+
+
+_GENERIC_TYPES_CACHE = GenericTypesCache()
+
+
+def _union_orderings_key(typevar_values: Any) -> Any:
+    """
+    This is intended to help differentiate between Union types with the same arguments in different order.
+    """
+    if isinstance(typevar_values, tuple):
+        args_data = []
+        for value in typevar_values:
+            args_data.append(_union_orderings_key(value))
+        return tuple(args_data)
+    elif typing_extensions.get_origin(typevar_values) is typing.Union:
+        return get_args(typevar_values)
+    else:
+        return ()
+
+
+def _early_cache_key(cls: type[BaseModel], typevar_values: Any) -> GenericTypesCacheKey:
+    """
+    This is intended for minimal computational overhead during lookups of cached types.
+
+    Note that this is overly simplistic, and it's possible that two different cls/typevar_values
+    inputs would ultimately result in the same type being created in BaseModel.__class_getitem__.
+    To handle this, we have a fallback _late_cache_key that is checked later if the _early_cache_key
+    lookup fails, and should result in a cache hit _precisely_ when the inputs to __class_getitem__
+    would result in the same type.
+    """
+    return cls, typevar_values, _union_orderings_key(typevar_values)
+
+
+def _late_cache_key(origin: type[BaseModel], args: tuple[Any, ...], typevar_values: Any) -> GenericTypesCacheKey:
+    """
+    This is intended for use later in the process of creating a new type, when we have more information
+    about the exact args that will be passed. If it turns out that a different set of inputs to
+    __class_getitem__ resulted in the same inputs to the generic type creation process, we can still
+    return the cached type, and update the cache with the _early_cache_key as well.
+    """
+    # Put the _union_orderings_key at the start here to ensure there cannot be a collision with an _early_cache_key
+    return _union_orderings_key(typevar_values), origin, args
+
+
+def set_cached_generic_type(
+    parent: type[BaseModel],
+    typevar_values: tuple[Any, ...],
+    type_: type[BaseModel],
+    origin: type[BaseModel] | None = None,
+    args: tuple[Any, ...] | None = None,
+) -> None:
+    _GENERIC_TYPES_CACHE[_early_cache_key(parent, typevar_values)] = type_
+    if len(typevar_values) == 1:
+        _GENERIC_TYPES_CACHE[_early_cache_key(parent, typevar_values[0])] = type_
+    if origin and args:
+        _GENERIC_TYPES_CACHE[_late_cache_key(origin, args, typevar_values)] = type_
+
+
+def get_cached_generic_type_early(parent: type[BaseModel], typevar_values: Any) -> type[BaseModel] | None:
+    return _GENERIC_TYPES_CACHE.get(_early_cache_key(parent, typevar_values))
+
+
+def get_cached_generic_type_late(
```

Comment:
If you find this crazy _early and _late cache key business hard to understand, I can explain in more detail (probably makes sense to add comments to these functions but wanted to see if you ended up wanting to simplify it first).

This approach was necessary to get `Model[List[T]][int]` to be the same type as `Model[List[int]]`. It could be modified to not use two different cache keys at different points, but the `_early_cache_key` is optimized to be as quick to compute as possible (for repeated-access speed), and the `_late_cache_key` is optimized to be as "correct" as possible, so that two types that are ultimately the same after resolving the type arguments will always produce cache hits.

Reply:
happy with two, let's add this to explanation to a docstring.